### PR TITLE
Add Kuryr SDN documentation.

### PR DESCRIPTION
**Pull request by dulek**
This PR adds documentation for recently merged OpenStack Kuryr support (https://github.com/openshift/openshift-ansible/pull/5574).

### CODE REVIEW

File: install_config/configuring_kuryrsdn.adoc [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+toc::[]
+
+[[kuryr-sdn-and-openshift]]
+== Kuryr SDN and {product-title}
+
+link:https://docs.openstack.org/kuryr-kubernetes/latest/[Kuryr] (or more
+specifically Kuryr-Kubernetes) is an SDN solution built using
+link:https://github.com/containernetworking/cni[CNI] and
+link:https://docs.openstack.org/neutron/latest/[OpenStack Neutron]. Its advantages
+include being able to use wide range of Neutron SDN backends and providing
```
Comment (celebdor) [inline install_config/configuring_kuryrsdn.adoc]: s/use wide range/use a wide range/

File: install_config/configuring_kuryrsdn.adoc [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+interconnectivity between Kubernetes pods and OpenStack VMs.
+
+Kuryr-Kubernetes and {product-title} integration is primarily designed for
+OpenStack virtual machines (VMs) installation. Kuryr-Kubernetes components are
```
Comment (celebdor) [inline install_config/configuring_kuryrsdn.adoc]: s/OpenStack virtual machines (VMs) installation/{product-title} clusters running on OpenStack virtual machines (VMs)/

File: install_config/configuring_kuryrsdn.adoc [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+installed as pods on {product-title}:
+
+* kuryr-controller - a single service instance, installed on any node. Modelled
+  in {product-title} as Deployment.
```
Comment (celebdor) [inline install_config/configuring_kuryrsdn.adoc]: s/as Depl/as a Depl/

File: install_config/configuring_kuryrsdn.adoc [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+* kuryr-cni - container installing and configuring Kuryr as CNI driver on each
+  {product-title} node. Modelled in {product-title} as DaemonSet.
```
Comment (celebdor) [inline install_config/configuring_kuryrsdn.adoc]: s/as Daem/as a Daem/

File: install_config/configuring_kuryrsdn.adoc [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+ os_sdn_network_plugin_name=cni
+
+ # Set userspace so that there's no iptables remains
+ openshift_node_proxy_mode='userspace'
+
+ # Disable management of the OpenShift Registry
+ openshift_hosted_manage_registry=false
+
+ # Disable management of the OpenShift Router
+ openshift_hosted_manage_router=false
```
Comment (celebdor) [inline install_config/configuring_kuryrsdn.adoc]: There should be an explanation about why it is not being deployed

File: install_config/configuring_kuryrsdn.adoc [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+ # Enable Kuryr
+ openshift_use_openshift_sdn=False
+ openshift_use_kuryr=True
+ os_sdn_network_plugin_name=cni
+
+ # Set userspace so that there's no iptables remains
+ openshift_node_proxy_mode='userspace'
+
+ # Disable management of the OpenShift Registry
+ openshift_hosted_manage_registry=false
```
Comment (celebdor) [inline install_config/configuring_kuryrsdn.adoc]: There should be an explanation about why it is not being deployed

File: install_config/configuring_kuryrsdn.adoc [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+ # Disable management of the OpenShift Router
+ openshift_hosted_manage_router=false
+
+ # Openstack access details
+ kuryr_openstack_auth_url=http://127.0.0.1/identity
+ kuryr_openstack_user_domain_name=default
+ kuryr_openstack_user_project_name=admin
+ kuryr_openstack_project_id=ec0b31802fd043c08bc15b74d2f9a3d3
+ kuryr_openstack_username=admin
+ kuryr_openstack_password=password
```
Comment (celebdor) [inline install_config/configuring_kuryrsdn.adoc]: These options above should have explanations

File: install_config/configuring_kuryrsdn.adoc [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+ # Neutron resources IDs
+ kuryr_openstack_pod_sg_id=f74c83a8-a520-421a-930e-21b6cd098c6a,01f85594-9950-4ded-a92c-5ad546a41188
+ kuryr_openstack_pod_subnet_id=c85cdee6-0ed1-4d8f-ae61-7afa4674b311
+ kuryr_openstack_pod_project_id=ec0b31802fd043c08bc15b74d2f9a3d3
+ kuryr_openstack_worker_nodes_subnet_id=477cfa49-e641-4d31-a7b5-5bc834743f61
+ kuryr_openstack_service_subnet_id=3b31a106-4084-4db9-bc0c-00b97afe186e
```
Comment (celebdor) [inline install_config/configuring_kuryrsdn.adoc]: Same here, these should be explained.

**Comment by dulek** [comment]
I've rebased the patch.

**Review by bogdando — APPROVED** [review]

File: _topic_map.yml [hunk 9]
```diff
@@ -645,6 +648,8 @@ Topics:
 - Name: Labeling Clusters for Amazon Web Services
   File: aws_cluster_labeling
   Distros: openshift-origin,openshift-enterprise
+- Name: Kuryr SDN Administration
```
Comment (ahardin-rh) [inline _topic_map.yml]: Is the Kuryr SDN Administration topic also for Origin only?

File: admin_guide/kuryr.adoc [hunk 10]
```diff
@@ -0,0 +1,32 @@
+[[admin-guide-kuryr]]
+= Kuryr SDN administration
```
Comment (ahardin-rh) [inline admin_guide/kuryr.adoc]: Administration (capitalize)

File: admin_guide/kuryr.adoc [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+:toc: macro
+:toc-title:
+:prewrap!:
+
+toc::[]
+
+== Overview
+xref:../../install_config/configuring_kuryrsdn.adoc#install-config-configuring-kuryr-sdn[Kuryr]
+(or Kuryr-Kubernetes) is one of the SDN choices for {product-title}. Kuryr uses
+OpenStack networking service - Neutron to connect pods to the network. This way
```
Comment (ahardin-rh) [inline admin_guide/kuryr.adoc]: Kuryr uses an OpenStack networking service, Neutron, to connect pods to the network. With this method, pods can have interconnectivity with OpenStack virtual machines (VMs), which is useful for {product-title} clusters deployed on OpenStack VMs.

File: admin_guide/kuryr.adoc [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+pods can have interconnectivity with OpenStack VMs, which is useful for
+{product-title} clusters deployed on OpenStack VMs.
+
+[[admin-guide-kuryr-orphaned-resources]]
+=== Orphaned OpenStack resources
```
Comment (ahardin-rh) [inline admin_guide/kuryr.adoc]: Orphaned OpenStack Resources

File: admin_guide/kuryr.adoc [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+All OpenStack resources created by Kuryr are tied to {product-title} resources
```
Comment (ahardin-rh) [inline admin_guide/kuryr.adoc]: tied to the the {product-title} resources lifecycle

File: admin_guide/kuryr.adoc [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+lifecycle. This means that if you delete your {product-title} cluster in a hard
+way (e.g. by deleting OpenStack VMs the cluster is running on), you may end up
```
Comment (ahardin-rh) [inline admin_guide/kuryr.adoc]: (that is, by deleting

File: admin_guide/kuryr.adoc [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+with orphaned resources on your OpenStack deployment. In particular this
+includes Neutron ports and Octavia (or LBaaS v2) load balancers, as well as
+networks, subnets and security groups precreated for Kuryr's use.
```
Comment (ahardin-rh) [inline admin_guide/kuryr.adoc]: subnets, and

File: architecture/topics/kuryr.adoc [hunk 16]
```diff
@@ -0,0 +1,47 @@
+xref:../../install_config/configuring_kuryrsdn.adoc#install-config-configuring-kuryr-sdn[Kuryr]
+(or more specifically Kuryr-Kubernetes) is an SDN solution built using
+link:https://github.com/containernetworking/cni[CNI] and
+link:https://docs.openstack.org/neutron/latest/[OpenStack Neutron]. Its
+advantages include being able to use a wide range of Neutron SDN backends and
+providing interconnectivity between Kubernetes pods and OpenStack VMs.
```
Comment (ahardin-rh) [inline architecture/topics/kuryr.adoc]: OpenStack virtual machines (VMs)

File: architecture/topics/kuryr.adoc [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+
+Kuryr-Kubernetes and {product-title} integration is primarily designed for
+{product-title} clusters running on OpenStack virtual machines (VMs).
```
Comment (ahardin-rh) [inline architecture/topics/kuryr.adoc]: OpenStack VMs.

File: architecture/topics/kuryr.adoc [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+
+[[architecture-additional-concepts-kuryr-openstack]]
+=== OpenStack deployment requirements
```
Comment (ahardin-rh) [inline architecture/topics/kuryr.adoc]: OpenStack Deployment Requirements

File: architecture/topics/kuryr.adoc [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+* Minimal service set is Keystone and Neutron.
+* A LBaaS service is required. It can be either
+  link:https://docs.openstack.org/octavia/latest/[Octavia] (with all of its own
+  dependencies) or deprecated Neutron-LBaaS v2.
+
+[[architecture-additional-concepts-kuryr-controller]]
+=== kuryr-controller
+
+kuryr-controller is a service responsible for watching {product-title} API for
+new pods being spawned and creating Neutron resources for them. For example when
```
Comment (ahardin-rh) [inline architecture/topics/kuryr.adoc]: For example, when

File: architecture/topics/kuryr.adoc [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+a pod gets created, kuryr-controller will notice that and call OpenStack Neutron
+to create a new port. Then information about that port (or VIF) is saved into
```
Comment (ahardin-rh) [inline architecture/topics/kuryr.adoc]: Then, information

File: architecture/topics/kuryr.adoc [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+pod's annotations. kuryr-controller is also able to use precreated port pools
```
Comment (ahardin-rh) [inline architecture/topics/kuryr.adoc]: saved into the pod's annotations.

File: architecture/topics/kuryr.adoc [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+for faster pod creation.
+
+Currently kuryr-controller must be run as a single service instance, so it's
```
Comment (ahardin-rh) [inline architecture/topics/kuryr.adoc]: Currently, kuryr-controller must

File: architecture/topics/kuryr.adoc [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+modeled in {product-title} as `Deployment` with `replicas=1`. It requires
```
Comment (ahardin-rh) [inline architecture/topics/kuryr.adoc]: , so it is modeled in

File: architecture/topics/kuryr.adoc [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+access to the underlying OpenStack service APIs.
+
+[[architecture-additional-concepts-kuryr-cni]]
+=== kuryr-cni
+
+kuryr-cni is (currently) a dummy container that is responsible for installing
+and configuring Kuryr CNI plugin on {product-title} nodes. It should be running
```
Comment (ahardin-rh) [inline architecture/topics/kuryr.adoc]: plug-in

File: architecture/topics/kuryr.adoc [hunk 25]
```diff
@@ -0,0 +1,47 @@
...
+on every node, so it's modelled in {product-title} as `DaemonSet`.
```
Comment (ahardin-rh) [inline architecture/topics/kuryr.adoc]: , so it is modeled in

File: architecture/topics/kuryr.adoc [hunk 26]
```diff
@@ -0,0 +1,47 @@
...
+
+Kuryr CNI plugin is called every time a new container that needs networking is
```
Comment (ahardin-rh) [inline architecture/topics/kuryr.adoc]: The Kuryr CNI plug-in

File: architecture/topics/kuryr.adoc [hunk 27]
```diff
@@ -0,0 +1,47 @@
...
+being spawned on the host. When that happens CNI plugin starts watching
```
Comment (ahardin-rh) [inline architecture/topics/kuryr.adoc]: is spawned on the host

Comment (ahardin-rh) [reply architecture/topics/kuryr.adoc]: When that happens, the CNI plug-in

File: architecture/topics/kuryr.adoc [hunk 28]
```diff
@@ -0,0 +1,47 @@
...
+{product-title} API to wait for kuryr-controller providing VIF annotation. When
+annotation is noticed, CNI plugin uses this info to plug the VIF to the
```
Comment (ahardin-rh) [inline architecture/topics/kuryr.adoc]: the CNI plug-in

File: install_config/configuring_kuryrsdn.adoc [hunk 29]
```diff
@@ -0,0 +1,106 @@
...
+
+[[kuryr-sdn-and-openshift]]
+== Kuryr SDN and {product-title}
+
+link:https://docs.openstack.org/kuryr-kubernetes/latest/[Kuryr] (or more
+specifically Kuryr-Kubernetes) is an SDN solution built using
+link:https://github.com/containernetworking/cni[CNI] and
+link:https://docs.openstack.org/neutron/latest/[OpenStack Neutron]. Its
+advantages include being able to use a wide range of Neutron SDN backends and
+providing interconnectivity between Kubernetes pods and OpenStack VMs.
```
Comment (ahardin-rh) [inline install_config/configuring_kuryrsdn.adoc]: OpenStack virtual machines (VMs).

File: install_config/configuring_kuryrsdn.adoc [hunk 30]
```diff
@@ -0,0 +1,106 @@
...
+
+Kuryr-Kubernetes and {product-title} integration is primarily designed for
+{product-title} clusters running on OpenStack virtual machines (VMs).
```
Comment (ahardin-rh) [inline install_config/configuring_kuryrsdn.adoc]: OpenStack VMs.

File: install_config/configuring_kuryrsdn.adoc [hunk 31]
```diff
@@ -0,0 +1,106 @@
...
+Kuryr-Kubernetes components are installed as pods on {product-title} in the
+`kube-system` namespace:
+
+* kuryr-controller - a single service instance, installed on any node. Modelled
```
Comment (ahardin-rh) [inline install_config/configuring_kuryrsdn.adoc]: Modeled

File: install_config/configuring_kuryrsdn.adoc [hunk 32]
```diff
@@ -0,0 +1,106 @@
...
+  in {product-title} as a `Deployment`.
+* kuryr-cni - container installing and configuring Kuryr as CNI driver on each
+  {product-title} node. Modelled in {product-title} as a `DaemonSet`.
```
Comment (ahardin-rh) [inline install_config/configuring_kuryrsdn.adoc]: Modeled

File: install_config/configuring_kuryrsdn.adoc [hunk 33]
```diff
@@ -0,0 +1,106 @@
...
+ kube-system   kuryr-cni-ds-77hkc                  1/1       Running   0          12m       10.11.0.5    10.11.0.5
+ kube-system   kuryr-cni-ds-9v8zs                  1/1       Running   0          12m       10.11.0.14   10.11.0.14
+ kube-system   kuryr-cni-ds-f9k0b                  1/1       Running   0          12m       10.11.0.11   10.11.0.11
+ kube-system   kuryr-cni-ds-pxthn                  1/1       Running   0          12m       10.11.0.8    10.11.0.8
+ kube-system   kuryr-cni-ds-t1m0s                  1/1       Running   0          12m       10.11.0.10   10.11.0.10
+ kube-system   kuryr-controller-4121192376-1vl1k   1/1       Running   0          1h        10.11.0.8    10.11.0.8
+----
+
+kuryr-cni pods should run on every {product-title} node. Single kuryr-controller
+deployment should run on any of the nodes.
```
Comment (ahardin-rh) [inline install_config/configuring_kuryrsdn.adoc]: deployments

File: admin_guide/kuryr.adoc [hunk 34]
```diff
@@ -0,0 +1,32 @@
...
+
+To get rid of them you can either look them up by ids that were specified in
+`kuryr.conf` or if you've created a separate OpenStack user for kuryr resources,
```
Comment (ahardin-rh) [inline admin_guide/kuryr.adoc]: Kuryr resources

File: install_config/configuring_kuryrsdn.adoc (hunk 33 above)
Comment (dulek) [reply install_config/configuring_kuryrsdn.adoc]: It haven't sounded right to me, I've changed it to:

 Single kuryr-controller instance should run on any of the nodes.

**Comment by dulek** [comment]
@ahardin-rh: Thank you for diligent review! I think I've addressed all of your comments.

File: _topic_map.yml [hunk 35]
```diff
@@ -642,7 +645,9 @@ Topics:
   File: idling_applications
 - Name: Analyzing Cluster Capacity
   File: cluster_capacity
-  Distros: openshift-origin,openshift-enterprise
+  Distros: openshift-origin
```
Comment (ahardin-rh) [inline _topic_map.yml]: @dulek All of your edits look great. I just have a question about line 648. This edit makes the Analyzing Cluster Capacity topic appear in Origin only, when it previously appeared in Origin and Enterprise docs. I think you meant to add a line after File :kuryr to define the distro for the Kuryr SDN Administration. So, I think this should be:

Name: Kuryr SDN Administration
File: kuryr
Distros: openshift-origin 

Please confirm, then we'll be ready to go. Thanks again!

Comment (dulek) [reply _topic_map.yml]: @ahardin-rh : My bad, thanks for noticing. It should be fine now.

**Comment by ahardin-rh** [comment]
@dulek Thank you!

**Comment by ahardin-rh** [comment]
Note to docs: This is Origin only and, therefore, no revision history is needed.

**Pull request merged**